File: frontend/types/agent.ts
/**
 * Agent 工作台类型定义
 *
 * 对应后端 app/schemas/agent.py 与 app/services/agent/progress.py 的事件协议。
 * 设计来源：2026-07-18-agent-functional-design.md §4/§6
 */

// ========== 会话与任务 ==========

export type SessionStatus = 'active' | 'archived' | 'deleted';

export type TaskStatus =
  | 'pending'
  | 'planning'
  | 'running'
  | 'awaiting_confirmation'
  | 'completed'
  | 'failed'
  | 'cancelled';

export const TERMINAL_TASK_STATUSES: ReadonlySet<TaskStatus> = new Set([
  'completed',
  'failed',
  'cancelled',
]);

export interface AgentSession {
  id: string;
  user_id: string;
  project_id?: string;
  title: string;
  status: SessionStatus;
  message_count: number;
  last_message_at?: string;
  created_at: string;
  updated_at: string;
}

export interface TokenUsage {
  prompt: number;
  completion: number;
  total: number;
}

export interface AgentTask {
  id: string;
  session_id: string;
  user_id: string;
  project_id?: string;
  query: string;
  plan?: Plan;
  status: TaskStatus;
  result?: {
    answer?: string;
    steps?: Array<Record<string, unknown>>;
    token_usage?: TokenUsage;
    cost_usd?: number;
    duration_sec?: number;
  };
  error?: string;
  started_at?: string;
  completed_at?: string;
  token_usage?: TokenUsage;
  cost_usd?: number;
}

// ========== 任务规划 ==========

export interface PlanStep {
  id: string;
  tool: string;
  args: Record<string, unknown>;
  depends_on: string[];
  description?: string;
}

export interface Plan {
  steps: PlanStep[];
  parallel_layers: string[][];
  reasoning?: string;
}

// ========== 工具调用 ==========

export interface ToolCall {
  step: number;
  tool: string;
  args: Record<string, unknown>;
  thought?: string;
}

export interface ToolResult {
  step: number;
  tool: string;
  success: boolean;
  data?: unknown;
  error?: string;
  duration_ms?: number;
  /** 缓存命中标记（后端从 LLM 缓存或工具结果缓存复用时返回） */
  cache_hit?: boolean;
  /** 服务端渲染提示：{type: chart|table|stats, payload} */
  display?: { type?: string; payload?: unknown };
}

export interface ToolInfo {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
  side_effects: boolean;
  required_role: string;
}

// ========== WebSocket 事件协议 ==========

export type WSEventType =
  | 'connected'
  | 'task_started'
  | 'plan'
  | 'thought'
  | 'tool_call'
  | 'tool_result'
  | 'confirmation_required'
  | 'final_response'
  | 'error'
  | 'task_completed'
  | 'task_cancelled'
  | 'token'
  | 'ping'
  | 'pong'
  | 'progress_snapshot'
  | 'not_found'
  | 'force_close'
  | 'unsubscribed';

export interface WSEvent<T = unknown> {
  type: WSEventType;
  task_id: string;
  timestamp: string;
  payload: T;
}

// 各类事件的 payload 类型
export interface ThoughtPayload {
  thought: string;
  step: number;
  max_steps: number;
}

/** 流式 token 事件 payload（LLM 边生成边推送） */
export interface TokenPayload {
  token: string;
  step?: number;
}

export interface PlanPayload extends Plan {}

export interface ToolCallPayload extends ToolCall {}

export interface ToolResultPayload extends ToolResult {}

export interface FinalResponsePayload {
  answer: string;
  references?: Array<{ title?: string; source?: string }>;
}

export interface ConfirmationRequiredPayload {
  task_id: string;
  tool: string;
  args: Record<string, unknown>;
}

export interface ErrorPayload {
  error: string;
  error_code?: string;
}

export interface TaskCompletedPayload {
  answer?: string;
  plan?: Plan;
  steps?: Array<Record<string, unknown>>;
  token_usage?: TokenUsage;
  cost_usd?: number;
  duration_sec?: number;
  status: TaskStatus;
}

// ========== 客户端发送给服务端的消息 ==========

export type WSClientMessageType =
  | 'subscribe'
  | 'unsubscribe'
  | 'cancel'
  | 'confirm'
  | 'ping';

export interface WSClientMessage<T = unknown> {
  type: WSClientMessageType;
  task_id?: string;
  payload?: T;
}

// ========== 前端聚合的消息项（用于 MessageList 渲染） ==========

export interface AgentMessage {
  id: string;
  role: 'user' | 'assistant' | 'tool' | 'system';
  content: string;
  toolCalls?: ToolCall[];
  toolResults?: ToolResult[];
  thought?: string;
  step?: number;
  timestamp: string;
  /** 标记为流式响应（LLM 边生成边显示），结构化事件到达时会被清理 */
  isStreaming?: boolean;
  /** 任务元数据（仅 assistant 最终消息有） */
  meta?: {
    task_id?: string;
    status?: TaskStatus;
    token_usage?: TokenUsage;
    cost_usd?: number;
    duration_sec?: number;
  };
}

File: backend/app/schemas/agent.py
"""Agent 相关 Pydantic 模型 — 请求/响应/事件

设计来源：2026-07-18-agent-functional-design.md §6
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.agent_session import SessionStatus
from app.models.agent_task import TaskStatus
from app.models.sandbox_execution import SandboxStatus


# ========== 会话 ==========

class SessionCreate(BaseModel):
    """创建会话请求"""
    title: Optional[str] = Field(None, max_length=200, description="会话标题，默认'新会话'")
    project_id: Optional[UUID] = Field(None, description="关联项目 ID")


class SessionResponse(BaseModel):
    """会话响应"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    project_id: Optional[UUID] = None
    title: str
    status: str = SessionStatus.ACTIVE
    message_count: int = 0
    last_message_at: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    context: Optional[Dict[str, Any]] = None  # 仅详情接口返回


# ========== 任务 ==========

class ChatRequest(BaseModel):
    """发起 Agent 对话"""
    session_id: UUID = Field(..., description="会话 ID")
    message: str = Field(..., min_length=1, max_length=10000, description="用户消息")
    project_id: Optional[UUID] = Field(None, description="项目 ID（覆盖会话默认）")
    tier: str = Field("fast_screen", description="分析层级: fast_screen / deep_insight")


class PlanStepView(BaseModel):
    """规划步骤视图"""
    id: str
    tool: str
    args: Dict[str, Any] = Field(default_factory=dict)
    depends_on: List[str] = Field(default_factory=list)
    description: Optional[str] = None


class PlanView(BaseModel):
    """任务规划视图"""
    steps: List[PlanStepView] = Field(default_factory=list)
    parallel_layers: List[List[str]] = Field(default_factory=list)
    reasoning: Optional[str] = None


class ToolCallView(BaseModel):
    """工具调用视图（推送给前端）"""
    step: int
    tool: str
    args: Dict[str, Any] = Field(default_factory=dict)
    thought: Optional[str] = None


class ToolResultView(BaseModel):
    """工具结果视图"""
    step: int
    tool: str
    success: bool
    data: Any = None
    error: Optional[str] = None
    display: Optional[Dict[str, Any]] = None
    duration_ms: Optional[int] = None


class TokenUsageView(BaseModel):
    """Token 用量"""
    prompt: int = 0
    completion: int = 0
    total: int = 0


class TaskResponse(BaseModel):
    """任务响应"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    session_id: UUID
    user_id: UUID
    project_id: Optional[UUID] = None
    query: str
    status: str = TaskStatus.PENDING
    current_step: Optional[int] = None
    plan: Optional[PlanView] = None
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    token_usage: Optional[TokenUsageView] = None
    cost_usd: Optional[float] = None
    duration_sec: Optional[float] = None
    created_at: datetime


class ChatResponse(BaseModel):
    """发起对话后的同步响应（任务异步执行）"""
    task_id: UUID
    session_id: UUID
    status: str = TaskStatus.PENDING
    message: str = "任务已创建，请通过 WebSocket 订阅进度"


# ========== 工具 ==========

class ToolInfo(BaseModel):
    """工具信息（用于 GET /tools 列表）"""
    name: str
    description: str
    parameters: Dict[str, Any]
    side_effects: bool
    required_role: str


# ========== 副作用确认 ==========

class ConfirmationRequest(BaseModel):
    """需要用户确认的副作用操作"""
    task_id: UUID
    step: int
    tool: str
    args: Dict[str, Any]
    description: str
    risk_level: str = "medium"  # low / medium / high


class ConfirmationResponse(BaseModel):
    """用户确认结果"""
    approved: bool = Field(..., description="是否同意执行")
    reason: Optional[str] = Field(None, description="拒绝原因（可选）")


# ========== WebSocket 事件 ==========

class WSEvent(BaseModel):
    """WebSocket 推送事件信封

    type 取值：
    - task_started: 任务开始执行
    - plan: 规划完成，推送 DAG
    - thought: LLM 思考过程
    - tool_call: 工具调用开始
    - tool_result: 工具执行结果
    - final_response: 最终答案
    - error: 错误
    - confirmation_required: 需要用户确认副作用
    - task_completed: 任务完成
    """
    type: str
    task_id: str
    timestamp: str
    payload: Dict[str, Any] = Field(default_factory=dict)


class WSClientMessage(BaseModel):
    """WebSocket 客户端消息

    type 取值：
    - subscribe: 订阅任务进度
    - unsubscribe: 取消订阅
    - confirm: 副作用确认
    - cancel: 取消任务
    - ping: 心跳
    """
    type: str
    task_id: Optional[str] = None
    payload: Dict[str, Any] = Field(default_factory=dict)


# ========== 沙箱 ==========

class SandboxExecuteRequest(BaseModel):
    """代码执行请求"""
    code: str = Field(..., min_length=1, max_length=50000, description="待执行代码")
    language: str = Field("python", description="编程语言")
    stdin: Optional[str] = Field(None, max_length=10000, description="标准输入")
    task_id: Optional[UUID] = Field(None, description="关联任务 ID")


class SandboxExecuteResponse(BaseModel):
    """代码执行响应"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    status: str = SandboxStatus.QUEUED
    stdout: Optional[str] = None
    stderr: Optional[str] = None
    exit_code: Optional[int] = None
    duration_ms: Optional[int] = None
    memory_kb: Optional[int] = None
